### PR TITLE
English version of CBO documentation

### PR DESCRIPTION
**Pull request by pavelvelikhov**
### Changelog entry 

...

### Changelog category 

* Documentation (changelog entry is not required)

### Additional information

...

### CODE REVIEW

File: ydb/docs/en/core/concepts/optimizer.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+A query plan is a graph of operations, such as reading data from a source, filtering a data stream by a predicate, or more complex operations such as [JOIN](../yql/reference/syntax/join.md) и [GROUP BY](../yql/reference/syntax/group_by.md). Optimizers in YDB take an initial query plan as input and transform it into a more efficient plan that is equivalent to the initial one in terms of the returned result.
+
+## Rule-based Optimizer
+
+A significant part of the optimizations in YDB is applicable to almost any query plan, which eliminates the need to analyze alternative plans and their costs. The rule-based optimizer consists of a set of heuristic rules that are applied whenever possible. For example, it is useful to filter out data as early as possible in the execution plan for any query. Each optimizer rule comprises of a condition that triggers the rule and rewriting logic that is executed when the plan is applied. Rules are applied iteratively as long as any rule conditions match.
```
Comment (anton-bobkov) [inline ydb/docs/en/core/concepts/optimizer.md:9]:
```suggestion
A significant part of the optimizations in YDB is applicable to almost any query plan, which eliminates the need to analyze alternative plans and their costs. The rule-based optimizer consists of a set of heuristic rules that are applied whenever possible. For example, it is useful to filter out data as early as possible in the execution plan for any query. Each optimizer rule comprises a condition that triggers the rule and rewriting logic that is executed when the plan is applied. Rules are applied iteratively as long as any rule conditions match.
```

File: ydb/docs/en/core/concepts/optimizer.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+The cost optimizer consists of three main components:
+
+* Plan enumerator
+* Cost estimation function
+* Statistics module, which is used to estimate statistics for the cost function
+
+### Plan enumerator
+
+The current Cost Based Optimizer in {{ ydb-short-name }} enumerates all useful join trees, for which the join conditions are defined. We first build a join hypergraph, where the nodes are tables and edges are join conditions. Depending on how the original query was written, the join hypergraph may have quite different topologies, varying from simple chain-like graphs to complex cliques. The resulting topology of the join graph determines how many possible alternite plans need to be considered by the optimizer.
```
Comment (anton-bobkov) [inline ydb/docs/en/core/concepts/optimizer.md:23]:
```suggestion
The current Cost-Based Optimizer in {{ ydb-short-name }} enumerates all useful join trees, for which the join conditions are defined. We first build a join hypergraph, where the nodes are tables and edges are join conditions. Depending on how the original query was written, the join hypergraph may have quite different topologies, varying from simple chain-like graphs to complex cliques. The resulting topology of the join graph determines how many possible alternite plans need to be considered by the optimizer.
```

File: ydb/docs/en/core/concepts/optimizer.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+The topology greatly influences the number of alternative plans that the optimizer needs to consider. Therefore, the cost-based optimizer limits the number of joins that are compared by exhaustive search, depending on the topology of the original plan. The capabilities of exact optimization in {{ ydb-short-name }} are listed in the following table:
+
+| Topology | Number of supported joins |
+| -------- | ------------------------- |
+| Chain | 110 |
+| Star | 18 |
+| Clique | 15 |
+
+{{ ydb-short-name }} uses a modification of the [DPHyp](https://www.researchgate.net/publication/47862092_Dynamic_Programming_Strikes_Back) algorithm to search for the best join order. DPHyp is modern dynamic programming algorithm for query optimization that avoids enumerating unnecessary alternatives and allows you to optimize plans with `JOIN` operators, complex predicates, and even `GROUP BY` and `ORDER BY` operators.
```
Comment (anton-bobkov) [inline ydb/docs/en/core/concepts/optimizer.md:60]:
```suggestion
{{ ydb-short-name }} uses a modification of the [DPHyp](https://www.researchgate.net/publication/47862092_Dynamic_Programming_Strikes_Back) algorithm to search for the best join order. DPHyp is a modern dynamic programming algorithm for query optimization that avoids enumerating unnecessary alternatives and allows you to optimize plans with `JOIN` operators, complex predicates, and even `GROUP BY` and `ORDER BY` operators.
```

File: ydb/docs/en/core/concepts/optimizer.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+### Cost Estimation Function
+
+In order to compare plans, we need to estimate their costs. Cost Function estimates the time and resources it takes to execute an operation in {{ ydb-short-name }}. The main parameters of the cost function are estimates of the input data size for each operator and the size of its result. These estimates are made based on statistics collected from {{ ydb-short-name }} tables, as well as an analysis of the plan itself.
+
+### Statistics for the Cost Based Optimizer {#statistics}
+
+Cost Based Optimizer uses tables statistics and individual column statistics. {{ ydb-short-name }} collects and maintains statistics in the background. You can force statistics collection using the [ANALYZE](../yql/reference/syntax/analyze.md) command.
```
Comment (anton-bobkov) [inline ydb/docs/en/core/concepts/optimizer.md:68]:
```suggestion
Cost Based Optimizer uses table statistics and individual column statistics. {{ ydb-short-name }} collects and maintains statistics in the background. You can force statistics collection using the [ANALYZE](../yql/reference/syntax/analyze.md) command.
```

File: ydb/docs/en/core/concepts/optimizer.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+Current set of table statistics:
+
+* Number of records;
```
Comment (anton-bobkov) [inline ydb/docs/en/core/concepts/optimizer.md:72]:
```suggestion
* Number of records
```

File: ydb/docs/en/core/concepts/optimizer.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+* Table size in bytes
+
+Current set of column statistics:
+
+* [Count-min sketch](https://en.wikipedia.org/wiki/Count%E2%80%93min_sketch).
```
Comment (anton-bobkov) [inline ydb/docs/en/core/concepts/optimizer.md:77]:
```suggestion
* [Count-min sketch](https://en.wikipedia.org/wiki/Count%E2%80%93min_sketch)
```

File: ydb/docs/en/core/yql/reference/yql-core/syntax/_includes/pragma/ydb.md [hunk 7]
```diff
@@ -2,6 +2,16 @@
...
 ## YDB
 
+### `ydb.CostBasedOptimizationLevel` {#costbasedoptimizationlevel}
+
+| Level | Description |
+| ------- | ---------------------- |
+| 0 | Cost Based Optimizer is turned off |
+| 1 | Cost Based Optimizer is turned off, but estimates are computed and avaliable |
+| 2 | Cost Based Optimizer is turned on only for queries that include [column oriented tables](../../../../../concepts/glossary.md#column-oriented-table) |
+| 3 | Cost Based Optimizer is turned on for all queries, but LookupJoin is prefered for row oriented tables |
```
Comment (anton-bobkov) [inline ydb/docs/en/core/yql/reference/yql-core/syntax/_includes/pragma/ydb.md:12]:
```suggestion
| 3 | Cost Based Optimizer is turned on for all queries, but LookupJoin is preferred for row oriented tables |
```

File: ydb/docs/en/core/yql/reference/yql-core/syntax/analyze.md [hunk 8]
```diff
@@ -0,0 +1,16 @@
+# ANALYZE
+
+`ANALYZE` forces the collection of statistics for {{ ydb-short-name }} Cost Based Optimizer.
```
Comment (anton-bobkov) [inline ydb/docs/en/core/yql/reference/yql-core/syntax/analyze.md:3]:
```suggestion
`ANALYZE` forces the collection of statistics for {{ ydb-short-name }} Cost-Based Optimizer.
```

File: ydb/docs/en/core/yql/reference/yql-core/syntax/analyze.md [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+Syntax:
+
+```yql
+ANALYZE <path_to_table> [ (<column_name> [, ...]) ]
+```
+
+This command forces synchronous collection of table statistics and column statistics for the specied columns, or all columns of the table if the columns were not specified. ANALYZE will return when all the specified statistics has been collect and is up to date.
```
Comment (anton-bobkov) [inline ydb/docs/en/core/yql/reference/yql-core/syntax/analyze.md:11]:
```suggestion
This command forces synchronous collection of table statistics and column statistics for the specied columns, or all columns of the table if the columns were not specified. ANALYZE will return when all the specified statistics has been collected and is up to date.
```

File: ydb/docs/en/core/yql/reference/yql-core/syntax/analyze.md [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
+* `path_to_table` — path to the table for which the statistis should be collected
```
Comment (anton-bobkov) [inline ydb/docs/en/core/yql/reference/yql-core/syntax/analyze.md:13]:
```suggestion
* `path_to_table` — path to the table for which the statistics should be collected
```

File: ydb/docs/en/core/concepts/optimizer.md [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+* Plan enumerator
+* Cost estimation function
+* Statistics module, which is used to estimate statistics for the cost function
+
+### Plan enumerator
+
+The current Cost-Based Optimizer in {{ ydb-short-name }} enumerates all useful join trees, for which the join conditions are defined. We first build a join hypergraph, where the nodes are tables and edges are join conditions. Depending on how the original query was written, the join hypergraph may have quite different topologies, varying from simple chain-like graphs to complex cliques. The resulting topology of the join graph determines how many possible alternite plans need to be considered by the optimizer.
+
+For example, a star is a common topology in analytical queries, where a main fact table is joined to multiple dimension tables:
```
Comment (blinkov) [inline ydb/docs/en/core/concepts/optimizer.md:25]:
```suggestion
### Plan enumerator

The current cost-based optimizer in {{ ydb-short-name }} enumerates all useful join trees for which the join conditions are defined. First, it builds a join hypergraph, where the nodes represent tables and the edges represent join conditions. Depending on how the original query is written, the join hypergraph can exhibit various topologies, ranging from simple chain-like graphs to complex cliques. The resulting topology of the join graph determines how many potential alternative plans the optimizer must consider.

For example, a star is a common topology in analytical queries, where a main fact table is joined to multiple dimension tables:
```

File: ydb/docs/en/core/concepts/optimizer.md [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+WHERE D.Year = 1997 AND  P.Product_Category = 'tv'
+
+GROUP BY
+    P.Brand,
+    S.Country
+```
+
+In this query graph, all `Dim...` tables are joined to the `Fact_Sales` fact table:
+![Join Graph](_assets/Star-Schema.png)
```
Comment (blinkov) [inline ydb/docs/en/core/concepts/optimizer.md:46]:
```suggestion
In this query graph, all `Dim...` tables are joined to the `Fact_Sales` fact table:

![Join graph](_assets/Star-Schema.png)
```

File: ydb/docs/en/core/concepts/optimizer.md [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+Common topologies also include chain and clique. A "chain" is a topology where tables are connected to each other sequentially and each table participates in no more than one join. A "clique" is a fully connected graph where each table is connected to another.
+
+In practice, OLAP queries often have a topology that is a combination of a "star" and a "chain", while complex topologies like a "clique" are very rare.
+
+The topology greatly influences the number of alternative plans that the optimizer needs to consider. Therefore, the cost-based optimizer limits the number of joins that are compared by exhaustive search, depending on the topology of the original plan. The capabilities of exact optimization in {{ ydb-short-name }} are listed in the following table:
```
Comment (blinkov) [inline ydb/docs/en/core/concepts/optimizer.md:52]:
```suggestion
Common topologies also include chains and cliques. A "chain" is a topology where tables are connected sequentially, with each table participating in no more than one join. A "clique" is a fully connected graph where each table is connected to every other table.

In practice, OLAP queries often exhibit a combination of "star" and "chain" topologies, while complex topologies like a "clique" are very rare.

The topology significantly impacts the number of alternative plans the optimizer must consider. Therefore, the cost-based optimizer limits the number of joins that are evaluated through an exhaustive search, depending on the topology of the original plan. The capabilities of exact optimization in {{ ydb-short-name }} are listed in the following table:
```

File: ydb/docs/en/core/concepts/optimizer.md [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+
+| Topology | Number of supported joins |
+| -------- | ------------------------- |
+| Chain | 110 |
+| Star | 18 |
+| Clique | 15 |
+
+{{ ydb-short-name }} uses a modification of the [DPHyp](https://www.researchgate.net/publication/47862092_Dynamic_Programming_Strikes_Back) algorithm to search for the best join order. DPHyp is a modern dynamic programming algorithm for query optimization that avoids enumerating unnecessary alternatives and allows you to optimize plans with `JOIN` operators, complex predicates, and even `GROUP BY` and `ORDER BY` operators.
```
Comment (blinkov) [inline ydb/docs/en/core/concepts/optimizer.md:60]:
```suggestion
{{ ydb-short-name }} uses a modification of the [DPHyp](https://www.researchgate.net/publication/47862092_Dynamic_Programming_Strikes_Back) algorithm to search for the optimal join order. DPHyp is a modern dynamic programming algorithm for query optimization that avoids enumerating unnecessary alternatives and enables the optimization of plans involving `JOIN` operators, complex predicates, and even `GROUP BY` and `ORDER BY` operators.
```

File: ydb/docs/en/core/concepts/optimizer.md [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+### Cost Estimation Function
+
+In order to compare plans, we need to estimate their costs. Cost Function estimates the time and resources it takes to execute an operation in {{ ydb-short-name }}. The main parameters of the cost function are estimates of the input data size for each operator and the size of its result. These estimates are made based on statistics collected from {{ ydb-short-name }} tables, as well as an analysis of the plan itself.
```
Comment (blinkov) [inline ydb/docs/en/core/concepts/optimizer.md:64]:
```suggestion
### Cost estimation function

To compare plans, the optimizer needs to estimate their costs. The cost function estimates the time and resources required to execute an operation in {{ ydb-short-name }}. The primary parameters of the cost function are estimates of the input data size for each operator and the size of its output. These estimates are based on statistics collected from {{ ydb-short-name }} tables, along with an analysis of the plan itself.
```

File: ydb/docs/en/core/concepts/optimizer.md [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+* Number of records
+* Table size in bytes
+
+Current set of column statistics:
+
+* [Count-min sketch](https://en.wikipedia.org/wiki/Count%E2%80%93min_sketch)
+
+Current cost optimization levels
+
+In {{ ydb-short-name }}, you can set the cost optimization level via pragma [CostBasedOptimizationLevel](../yql/reference/syntax/pragma.md#costbasedoptimizationlevel).
```
Comment (blinkov) [inline ydb/docs/en/core/concepts/optimizer.md:81]:
```suggestion
### Statistics for the cost-based optimizer {#statistics}

The cost-based optimizer relies on table statistics and individual column statistics. {{ ydb-short-name }} collects and maintains these statistics in the background. You can manually force statistics collection using the [ANALYZE](../yql/reference/syntax/analyze.md) query.

The current set of table statistics includes:

* Number of records
* Table size in bytes

The current set of column statistics includes:

* [Count-min sketch](https://en.wikipedia.org/wiki/Count%E2%80%93min_sketch)

### Cost optimization levels

In {{ ydb-short-name }}, you can configure the cost optimization level via the [CostBasedOptimizationLevel](../yql/reference/syntax/pragma.md#costbasedoptimizationlevel) pragma.
```

File: ydb/docs/en/core/concepts/toc_i.yaml [hunk 17]
```diff
@@ -23,4 +23,6 @@ items:
   - name: Disk subsystem of a cluster
     href: cluster/distributed_storage.md
 - name: Federated query
-  include: { path: federated_query/toc_p.yaml, mode: link }
\ No newline at end of file
+  include: { path: federated_query/toc_p.yaml, mode: link }
+- name: Query Optimizer
```
Comment (blinkov) [inline ydb/docs/en/core/concepts/toc_i.yaml:27]:
```suggestion
- name: Query optimizer
```

File: ydb/docs/en/core/yql/reference/yql-core/syntax/_includes/pragma/ydb.md [hunk 18]
```diff
@@ -2,6 +2,16 @@
...
 
+### `ydb.CostBasedOptimizationLevel` {#costbasedoptimizationlevel}
+
+| Level | Description |
+| ------- | ---------------------- |
+| 0 | Cost Based Optimizer is turned off |
+| 1 | Cost Based Optimizer is turned off, but estimates are computed and avaliable |
+| 2 | Cost Based Optimizer is turned on only for queries that include [column oriented tables](../../../../../concepts/glossary.md#column-oriented-table) |
+| 3 | Cost Based Optimizer is turned on for all queries, but LookupJoin is preferred for row oriented tables |
+| 4 | Cost Based Optimizer is turned on for all queries |
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/syntax/_includes/pragma/ydb.md:13]:
```suggestion
### `ydb.CostBasedOptimizationLevel` {#costbasedoptimizationlevel}

| Level | Description |
| ----- | ------------------------ |
| 0     | Cost-based optimizer is disabled. |
| 1     | Cost-based optimizer is disabled, but estimates are computed and available. |
| 2     | Cost-based optimizer is enabled only for queries that include [column-oriented tables](../../../../../concepts/glossary.md#column-oriented-table). |
| 3     | Cost-based optimizer is enabled for all queries, but `LookupJoin` is preferred for row-oriented tables. |
| 4     | Cost-based optimizer is enabled for all queries. |
```

File: ydb/docs/en/core/yql/reference/yql-core/syntax/analyze.md [hunk 19]
```diff
@@ -0,0 +1,16 @@
+# ANALYZE
+
+`ANALYZE` forces the collection of statistics for {{ ydb-short-name }} Cost-Based Optimizer.
+
+Syntax:
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/syntax/analyze.md:5]:
```suggestion
# ANALYZE

`ANALYZE` forces the collection of statistics for [{{ ydb-short-name }} cost-based optimizer](../../../concepts/optimizer.md).

## Syntax
```

File: ydb/docs/en/core/yql/reference/yql-core/syntax/analyze.md [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+```yql
+ANALYZE <path_to_table> [ (<column_name> [, ...]) ]
+```
+
+This command forces synchronous collection of table statistics and column statistics for the specied columns, or all columns of the table if the columns were not specified. ANALYZE will return when all the specified statistics has been collected and is up to date.
+
+* `path_to_table` — path to the table for which the statistics should be collected
+* `column_name` — collect column statistics only for specified columns of the table
+
+Current set of statistics is described here [{#T}](../../../concepts/optimizer.md#statistics).
```
Comment (blinkov) [inline ydb/docs/en/core/yql/reference/yql-core/syntax/analyze.md:16]:
```suggestion
This command forces the synchronous collection of table statistics and column statistics for the specified columns or for all columns if none are specified. `ANALYZE` returns once all the requested statistics have been collected and are up to date.

* `path_to_table` — the path to the table for which statistics should be collected.
* `column_name` — collect column statistics only for the specified columns of the table.

The current set of statistics is described in [{#T}](../../../concepts/optimizer.md#statistics).
```

**Review by blinkov — DISMISSED** [review]

File: ydb/docs/en/core/concepts/optimizer.md [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+
+## Cost-Based Query Optimizer
+
+Cost-Based Optimizer is used for more complex optimizations, such as choosing an optimal join order and join algorithms. Cost-based Optimizer considers a large number of alternative execution plans for each query and chooses the best one based on the cost estimate for each option. Currently, this optimizer only works with plans that contain [JOIN](../yql/reference/syntax/join.md) operations. It chooses the best order for these operations to be executed, and also chooses the most efficient implementation for each Join algoritm in the plan.
+
+The cost optimizer consists of three main components:
+
+* Plan enumerator
+* Cost estimation function
+* Statistics module, which is used to estimate statistics for the cost function
```
[the remark this answers is not in the document]
Comment (pavelvelikhov) [reply ydb/docs/en/core/concepts/optimizer.md:19]: We have defined Cost Based Optimizer, so we don't need to use a determiner before, it sounds much better w/o one

File: ydb/docs/en/core/concepts/optimizer.md (hunk 11 above)
Comment (pavelvelikhov) [reply ydb/docs/en/core/concepts/optimizer.md:25]: I don't see any reasoning why you think your version is better (except correcting capitalization and a typo)

File: ydb/docs/en/core/concepts/optimizer.md (hunk 13 above)
Comment (pavelvelikhov) [reply ydb/docs/en/core/concepts/optimizer.md:52]: I don't agree to these changes - especially 'evaluated though an' - this is more confusing, it gives an impression that we are evaluating joins - which is not the case

File: ydb/docs/en/core/concepts/optimizer.md (hunk 14 above)
Comment (pavelvelikhov) [reply ydb/docs/en/core/concepts/optimizer.md:60]: Here the changes are worse than the original - replacing simpler terms with passive constructs like 'enables the optimization of' or 'involving'

**Review by fomichev3000 — APPROVED** [review]

**Pull request merged**